util/gpool/spmc/spmcpool.go

```diff
@@ -0,0 +1,403 @@
+			break
+		}
+
+		p.lock.Lock()
+		expiredWorkers := p.workers.retrieveExpiry(p.options.ExpiryDuration)
+		p.lock.Unlock()
+
+		// Notify obsolete workers to stop.
+		// This notification must be outside the p.lock, since w.task
+		// may be blocking and may consume a lot of time if many workers
+		// are located on non-local CPUs.
+		for i := range expiredWorkers {
+			expiredWorkers[i].taskBoxCh <- nil
+			expiredWorkers[i] = nil
+		}
+
+		// There might be a situation where all workers have been cleaned up(no worker is running),
+		// or another case where the pool capacity has been Tuned up,
+		// while some invokers still get stuck in "p.cond.Wait()",
+		// then it ought to wake all those invokers.
+		if p.Running() == 0 || (p.Waiting() > 0 && p.Free() > 0) {
+			p.cond.Broadcast()
+		}
+	}
+}
+
+// Tune changes the capacity of this pool, note that it is noneffective to the infinite or pre-allocation pool.
+func (p *Pool[T, U, C, CT, TF]) Tune(size int) {
+	capacity := p.Cap()
+	if capacity == -1 || size <= 0 || size == capacity {
+		return
+	}
+	p.capacity.Store(int32(size))
+	if size > capacity {
+		// boost
+		if size-capacity == 1 {
+			p.cond.Signal()
+			return
+		}
+		p.cond.Broadcast()
+	}
+}
+
+// Running returns the number of workers currently running.
+func (p *Pool[T, U, C, CT, TF]) Running() int {
+	return int(p.running.Load())
+}
+
+// Free returns the number of available goroutines to work, -1 indicates this pool is unlimited.
+func (p *Pool[T, U, C, CT, TF]) Free() int {
+	c := p.Cap()
+	if c < 0 {
+		return -1
+	}
+	return c - p.Running()
+}
+
+// Waiting returns the number of tasks which are waiting be executed.
+func (p *Pool[T, U, C, CT, TF]) Waiting() int {
+	return int(p.waiting.Load())
+}
+
+// IsClosed indicates whether the pool is closed.
+func (p *Pool[T, U, C, CT, TF]) IsClosed() bool {
+	return p.state.Load() == gpool.CLOSED
+}
+
+// Cap returns the capacity of this pool.
+func (p *Pool[T, U, C, CT, TF]) Cap() int {
+	return int(p.capacity.Load())
+}
+
+func (p *Pool[T, U, C, CT, TF]) addRunning(delta int) {
+	p.running.Add(int32(delta))
+}
+
+func (p *Pool[T, U, C, CT, TF]) addWaiting(delta int) {
+	p.waiting.Add(int32(delta))
+}
+
+// release closes this pool and releases the worker queue.
+func (p *Pool[T, U, C, CT, TF]) release() {
+	if !p.state.CompareAndSwap(gpool.OPENED, gpool.CLOSED) {
+		return
+	}
+	p.lock.Lock()
+	p.workers.reset()
+	p.lock.Unlock()
+	// There might be some callers waiting in retrieveWorker(), so we need to wake them up to prevent
+	// those callers blocking infinitely.
+	p.cond.Broadcast()
+}
+
+func isClose(exitCh chan struct{}) bool {
+	select {
+	case <-exitCh:
+		return true
+	default:
+	}
+	return false
+}
+
+// ReleaseAndWait is like Release, it waits all workers to exit.
+func (p *Pool[T, U, C, CT, TF]) ReleaseAndWait() {
+	if p.IsClosed() || isClose(p.stopCh) {
+		return
+	}
+
+	close(p.stopCh)
+	p.release()
+	for {
+		if p.Running() == 0 && p.heartbeatDone.Load() {
```

Comment:
This is a super fast loop. Is it expected?

Reply:
Please address this comment

Reply:
> Please address this comment

Done

Reply:
> This is a super fast loop. Is it expected?

```release``` just tell workers to exit and we need to wait for them to exit in the loop. but it is fast to exit the loop.